### PR TITLE
[Minor] access intermediate validation result via draft

### PR DESCRIPTION
**Pull request by ealush**
| Q                | A
| ---------------- | ---
| Bug fix?         | no
| New feature?     | yes
| Breaking change? | no
| Deprecations?    | no
| Documentation?   | yes
| Tests added?     | yes

This new feature allows using the intermediate result from within the suite itself.

## Accessing the intermediate test result from within your suite
In some cases, you might want to access the intermediate result, for example, if you'd like to stop a server call from happening for a field that's already invalid. To do that, you can access the draft of the result object from within the test suite, using its second argument.

In the following example, we're preventing the async validation from running over the username field in case it already has errors.

```js
passable('NewUserForm', (test, draft) => {
    test('username', 'Must be a string between 2 and 10 chars', () => {
        enforce(data.username).isString().largerThan(1).smallerThan(11);
    });

    if (!draft.getErrors('username').length) {
        // if the username did not pass the previous test, the following test won't run
        test('username', 'already exists', fetch(`/check_availability?username=${data.username}`));
    }
});
```

### CODE REVIEW

File: src/Passable.js [hunk 1]
```diff
@@ -65,48 +65,60 @@ class Passable {
...
+            this.clearPendingTest(test);
+        }
+        this.res.bumpTestCounter(test.fieldName);
+    }
+
+    /**
+     * calls `runTest` on all pending tests, clears pending tests list and bumps counters
+     */
+    runPendingTests = () => {
+        [...this.pending].forEach(this.runTest);
```
Comment (raphaelboukara) [inline src/Passable.js]: superb very clear ❤️

File: src/Passable.js [hunk 2]
```diff
@@ -65,48 +65,60 @@ class Passable {
...
-        test.severity = severity;
-
-        this.addPendingTest(test);
+        if (test instanceof Promise) {
+            this.addPendingTest(Object.assign(test, {
+                fieldName,
+                statement,
+                severity
+            }));
+        }
```
Comment (raphaelboukara) [inline src/Passable.js]: what do you think about:

```js
test = (fieldName, statement, test, severity) => {

        if (this.specific.excludes(fieldName)) {
            this.res.addToSkipped(fieldName);
            return;
        }

        this.res.initFieldCounters(fieldName);

        this.addPendingTest(Promise.resolve().then(() => {
            return Object.assign(test, {
                fieldName,
                statement,
                severity
            });
        }
    }
```

File: src/Passable.js [hunk 3]
```diff
@@ -65,48 +65,60 @@ class Passable {
...
-                this.clearPendingTest(test);
+            if (!isValid) {
+                this.res.fail(test.fieldName, test.statement, test.severity);
             }
-            this.res.bumpTestCounter(test.fieldName);
-        });
+            this.clearPendingTest(test);
+        }
+        this.res.bumpTestCounter(test.fieldName);
+    }
```
Comment (raphaelboukara) [inline src/Passable.js]: and here:
```js
runTest = (test) => {
        this.res.markAsync(test.fieldName);

        const done = () => {
            this.res.markAsDone(test.fieldName);
            this.clearPendingTest(test);
        };

        const fail = () => {
            // order is important here! fail needs to be called before `done`.
            this.res.fail(test.fieldName, test.statement, test.severity);
            done();
        };

        testRunnerAsync(test, done, fail);

        this.res.bumpTestCounter(test.fieldName);
    }
```

File: src/Passable.js [hunk 4]
```diff
@@ -65,48 +65,60 @@ class Passable {
 
         this.res.initFieldCounters(fieldName);
 
-        if (typeof test !== 'function' && !(test instanceof Promise)) {
-            return;
+        if (typeof test === 'function') {
+            this.runTest(Object.assign(test, {
```
Comment (sahariko) [inline src/Passable.js]: You're merging the objects anyhow, so how about:
```js
let operation;

if (typeof test === 'function') {
    operation = 'runTest';
} else if (test instanceof Promise) {
    operation = 'addPendingTest';
} else {
    return;
}

Object.assign(test, {
    fieldName,
    statement,
    severity
});

this[operation](test);
```

Comment (ealush) [reply src/Passable.js]: Nice!

**Pull request merged**